amethyst_assets/src/storage.rs

```diff
@@ -130,120 +142,145 @@ impl<A: Asset> AssetStorage<A> {
-                            reloads.push((handle.downgrade(), old_reload));
-
-                            continue;
+                    Processed::HotReload {
+                        data,
+                        handle,
+                        name,
+                        old_reload,
+                    } => {
+                        let (asset, reload_obj) = match data.map(|FormatValue { data, reload }| {
+                            (data, reload)
+                        }).and_then(|(d, rel)| f(d).map(|a| (a, rel)))
+                            .chain_err(|| ErrorKind::Asset(name.clone()))
+                        {
+                            Ok((ProcessingState::Loaded(x), r)) => (x, r),
+                            Ok(..) => {
+                                continue; // FIXME
+                            }
+                            Err(e) => {
+                                error!(
+                                    "{:?}: Failed to hot-reload asset {:?} (handle id: {:?}): {}\n\
+                                     Falling back to old reload object.",
+                                    A::NAME,
+                                    name,
+                                    handle,
+                                    e,
+                                );
+
+                                reloads.push((handle.downgrade(), old_reload));
+
+                                continue;
+                            }
+                        };
+
+                        let id = handle.id();
+                        assert!(
+                            bitset.contains(id),
+                            "Expected handle {:?} to be valid, but the asset storage says otherwise",
+                            handle,
+                        );
+                        unsafe {
+                            let old = assets.get_mut(id);
+                            *old = asset;
                         }
-                    };
 
-                    let id = handle.id();
-                    assert!(
-                        bitset.contains(id),
-                        "Expected handle {:?} to be valid, but the asset storage says otherwise",
-                        handle,
-                    );
-                    unsafe {
-                        let old = assets.get_mut(id);
-                        *old = asset;
+                        (reload_obj, handle)
                     }
+                };
 
-                    (reload_obj, handle)
+                // Add the reload obj if it is `Some`.
+                if let Some(reload_obj) = reload_obj {
+                    reloads.push((handle.downgrade(), reload_obj));
                 }
-            };
+            }
 
-            // Add the reload obj if it is `Some`.
-            if let Some(reload_obj) = reload_obj {
-                reloads.push((handle.downgrade(), reload_obj));
+            for p in requeue.drain(..) {
+                self.processed.push(p);
```

Comment:
I don't understand why you first push the elements to `requeue` just to `drain` them here..

Reply:
Because if I push them in msqueue I'll create an infinite loop